### ISSUE TITLE
[NEWS] Nowy zasób do dodawania numeru przesyłki

### ISSUE THREAD

**Issue by MartaNowaczyk** [allegro/allegro-api#1217]
Udostępniliśmy nowy zasób w wersji beta **[POST /order/checkout-forms/{checkoutForm.id}/shipments](https://developer.allegro.pl/documentation/#tag/Order-management/paths/~1order~1checkout-forms~1{id}~1shipments/post)**, dzięki któremu dodasz numer przesyłki do zamówienia. Możesz go przypisać do dowolnie wybranych lineItems.id z danego zamówienia.

Pracujemy też nad metodą GET do pobierania numerów przesyłek, które dodałeś do konkretnego zamówienia. Powinna być dostępna w najbliższych tygodniach. Postaramy się opublikować ją jak najwcześniej.

[Więcej…](https://developer.allegro.pl/orders/#08)

**Comment by kuli-g** [comment]
Dlaczego numer przesyłki dopisuje się po podaniu lineItems.id zamiast po prostu samym checkoutForm.id?
Jaki jest w końcu główny identyfikator zamówienia żeby nie trzeba było przy każdej nowej funkcji mapować na coś innego?

**Comment by Draken79** [comment]
czy jesli kilka przedmiotow wysylam jedną przesylka to wystarczy, ze podam lineItems.id tylko jednego z nich? Czy trzeba dodawac wszytskie lineItems.id?

Troche to bez sensu jednak, przychylam sie do pytania kuli-g. Dlaczego nie można tego zrobić po checkoutForm.id - przecież w Allegro, nie można rozdzielić jednego zamówienia na cześci.

**Comment by MartaNowaczyk** [comment]
@kuli-g checkout-form jest unikatowy i jednoznaczny. Dlaczego lineItems.id wyjaśniam poniżej.

@Draken79 To jak sobie ograsz to po swojej stronie to już pozostawiam tobie. Odpowiadając na pytanie, tak możesz wpisać tylko jeden lineItems.id.

Poza tym rozdzielać zakupów nie można, ale można określić w ilu paczkach dane zamówienie zostanie wysłane w zależności od liczby zakupionych produktów. Jeśli do jednego zamówienia wysyłasz 3 paczki, wtedy możesz wybrać przy dodawaniu przesyłki odpowiednie lineItems.id

**Comment by kuli-g** [comment]
@MartaNowaczyk 99,9% zamówień realizowanych jest w jednej paczce ale wy oczywiście bardziej się przejmujecie tym 0,1% niż przejrzystością i prostotą. Teraz muszę do każdego zamówienia mapować jeszcze lineItems.id żeby dodać numer paczki do zamówienia który przecież ten lineItems.id zawiera. Jak dla mnie bez sensu, można by to zrobić jako "opcjonalne".

**Comment by MartaNowaczyk** [comment]
@kuli-g jeśli chodzi o to "opcjonalnie" to zgłosiłam już taką propozycję po waszych pierwszych wpisach.

**Comment by kuli-g** [comment]
@MartaNowaczyk wspaniale. To teraz dogadajcie się w końcu z inpostem żeby można było używać checkout-form bo już pół roku minęło od waszego "pracujemy nad tym" i będzie można się przełączać na restapi.

**Comment by MartaNowaczyk** [comment]
@kuli-g tak jak pisaliśmy w #1043 - w harmonogramie mamy to na Luty/Marzec 2019:

` wspólnie z InPostem rozwiążemy problem integracji związany z nowymi numerami zamówień.`

**Comment by alekskuc** [comment]
Witam, jakie operatorId mają dodatkowi operatorzy przesyłek "X_PRESS_COURIERS" oraz "RHENUS_LOGISTICS" żeby ich ustawić metodą WebAPI doAddPackageInfoToPostBuyForm?

**Comment by MartaNowaczyk** [comment]
@alekskuc nie mają, dostępne są tylko przez REST API

**Comment by krzysztofgwiazda** [comment]
@MartaNowaczyk w powyższym komunikacie sugerujecie Państwao iż zasób jest w wersji beta, natomiast w dokumentacji nie jest to oznaczone. Czy zatem jest to już funkcjonalność produkcyjna?

**Comment by MartaNowaczyk** [comment]
@krzysztofgwiazda zasób jest w wersji beta. Rzeczywiście brak takiego oznaczenia. Przy najbliższym wdrożeniu postaramy się je dodać. Dzięki za czujność.

**Comment by jakubmaguza** [comment]
Są jakieś informacje czy uda się dodawać numer przesyłki po CheckoutForms.id?

**Comment by MartaNowaczyk** [comment]
@jakubmaguza na ten moment nie mam nowych informacji. Jeśli coś się zmieni damy znać. Na ten moment opracowujemy pobieranie numerów przesyłek.

**Comment by neldh** [comment]
Podbijam temat. Dałoby się zrobić końcówkę do dodawania numeru przesylki po checkoutForm.id (najlepiej), surcharge.id lub payment.id?
Ta informacja byłaby niezmiernie cenna dla mnie w tej chwili, gdyż blokuje kolejny etap przechodzenia na rest api z webapi. Jeśli planujecie wdrożenie takiego czegoś, odpadałaby mi konieczność zaciągania checkout forma za każdym razem, gdy chce dodac nr przesyłki, tylko po to, zeby poznać id line itemów.
Wystarczyłoby w aktualnej koncowce dodać pole w stylu "allLineItems": true

**Comment by MartaNowaczyk** [comment]
@neldh jak już to po checkoutForm.id. Jeszcze nie ustaliliśmy co z tym zrobimy. Jak będę miała takie informacje dam znać.

**Comment by jakubmaguza** [comment]
Czy dla klienta jest jakakolwiek różnica czy podam jeden LineItemID czy wszystkie?

**Comment by MartaNowaczyk** [comment]
@jakubmaguza nie, nie powinno być różnicy.

**Comment by repmic** [comment]
Mam problem z dodaniem numeru przesyłki podczas wykonania POST na zasobie /order/checkout-forms/{checkoutForm.id}/shipments otrzymuję w odpowiedzi błąd {
    "errors": [
        {
            "code": "ERROR",
            "message": "An error has occurred",
            "details": null,
            "path": "/order/checkout-forms/3fd5b034-4359-11e9-ad58-73372effae00/shipments",
            "userMessage": null
        }
    ]
}
Nie wiem co może być tego powodem.
Proszę o pomoc.

**Comment by jakubmaguza** [comment]
Pokaż zapytanie, na pewno się przyda :)

**Comment by vendoremo** [comment]
Czy są plany, żeby LineItemID były opcjonalne w dodawaniu nr przesyłki? Albo żeby domyślnie polecenie dotyczyło wszystkich bez konieczności ich podawania? Poprzez stronę allegro nie da się dodać nr przesyłki osobno dla każdego produktu w zamówieniu, więc nie do końca rozumiem logikę umożliwienia takiej operacji poprzez api.

**Comment by neldh** [comment]
Jescze podbiję to:
Jeśli takie plany są, to kiedy to zostanie dodane?

**Comment by repmic** [comment]
@vendoremo curl -X POST \
    https://api.allegro.pl/order/checkout-forms/9ffbc480-43f0-11e9-be21-db769f8b6fe9/shipments \
    -H 'Authorization: Bearer {token}' \
    -H 'Content-Type: application/vnd.allegro.beta.v1+json' \
    -H 'Accept: application/vnd.allegro.beta.v1+json' 
    -d '{
    "carrierId": "DHL"                        
    "waybill": "12345678910PL",              
    "carrierName": null,                      
    "lineItems": [{"id": "9ffbc480-43f0-11e9-be21-db769f8b6fe9"}] 
}

**Comment by MartaNowaczyk** [comment]
@vendoremo w linijce "carrierId": "DHL" brakuje na końcu przecinka

**Comment by MartaNowaczyk** [comment]
@vendoremo @neldh na ten moment nie ma takich planów.

**Comment by neldh** [comment]
@repmic 
eee... Chodzi Ci o to, że jak w lineItems podam id checkout forma to to przejdzie? 0_o

@MartaNowaczyk 
W takim razie pytanie czy to tak zostanie, czy za miesiąc przestanie działać, bo "naprawicie" ?

**Comment by jakubmaguza** [comment]
@repmic 
GUID checkout forma i GUID lineItems są identyczne, a nie powinno tak być. To są dwa różne GUIDy. Są bardzo podobne, ale jednak różne :)

@MartaNowaczyk 
Dlaczego te guidy są aż tak do siebie zbliżone? Ma to jakiś cel, czy po prostu tak jest?

**Comment by aktywnitu** [comment]
Dzień dobry,

chciałem przejść na metodę z REST API, ale niestety otrzymuję błąd podczas wysyłki numerów:
```
 headers = {
            'Authorization': "Bearer {}".format(self.token),
            'Accept': 'application/vnd.allegro.beta.v1+json',
            'Content-Type': 'application/vnd.allegro.beta.v1+json',
        }
{'waybill': '523000013926025017120940', 'carrierId': 'INPOST', 'lineItems': [{'id': '304faaf0-44ce-11e9-a794-5bae4784f84c'}]}
https://api.allegro.pl/order/checkout-forms/304faaf2-44ce-11e9-a794-5bae4784f84c/shipments 400 <Response [400]>
{'errors': [{'code': 'ERROR', 'message': 'An error has occurred', 'path': '/order/checkout-forms/304faaf2-44ce-11e9-a794-5bae4784f84c/shipments', 'userMessage': None, 'details': None}]}
```

**Comment by repmic** [comment]
@MartaNowaczyk dodanie przecinka nic nie zmienia. W dokumentacji na stronie https://developer.allegro.pl/orders/#addTracking też nie ma tego przecinka.

**Comment by repmic** [comment]
@MartaNowaczyk jednak dodanie przecinka rozwiązało problem. Myślę że warto poprawić dokumentację.

**Comment by MartaNowaczyk** [comment]
@rempic tak, widziałam, że w poradniku go brakuje. Poprawimy przy kolejnym wdrożeniu na stronie.

**Comment by MartaNowaczyk** [comment]
@neldh na ten moment to wersja beta, więc możemy jeszcze coś zmienić.

@jakubmaguza tak po prostu jest

@aktywnitu w zapytaniu ' zamień na ", bo nic innego tu nie widzę jeśli chodzi o niepoprawne dane JSON

**Comment by repmic** [comment]
@MartaNowaczyk Drugą nieścisłością w poradniku jest opcja lineItems z tego co testowałem to również jest wymagana, więc warto dopisać jako wymagane w komentarz,y gdyż w pozostałych jest komentarz wymagane, natomiast jeżeli tu nie ma to nie dodawałem tego i odpowiedź sygnalizowała bład.

**Comment by MartaNowaczyk** [comment]
@repmic trafna uwaga, dodałam do zmian na stronie.

**Comment by aktywnitu** [comment]
> @aktywnitu w zapytaniu ' zamień na ", bo nic innego tu nie widzę jeśli chodzi o niepoprawne dane JSON
Błąd jest nadal. Próbując poprzez curla, zauważyłem, że z początku otrzymuję odpowiedź pozytywną z createdAt na przemian z błędem. A w tej chwili już tylko info o błędzie nic nie mówiącym:
`{"errors":[{"code":"ERROR","message":"An error has occurred","details":null,"path":"/order/checkout-forms/93b03802-41bb-11e9-a819-2b9105cf628d/shipments","userMessage":null}]}`
Czy nie można drugi raz wysłać numeru?

**Comment by aktywnitu** [comment]
> @aktywnitu w zapytaniu ' zamień na ", bo nic innego tu nie widzę jeśli chodzi o niepoprawne dane JSON

Ma Pani rację. Należało przesłać poprawny format JSONa.

**Comment by vendoremo** [comment]
Nie da się dodać nr przesyłki do zamówienia, dla którego wybrano formę dostawy list zwykły. Na pierwszy rzut oka jest to logiczne. Czasem jednak sprzedawca z różnych względów decyduje się wysłać towar listem poleconym. Czy możecie przywrócić możliwość dodania nr przesyłki do tego typu zamówień? To jest z korzyścią dla klienta, nie widzę powodu, żeby to blokować. Z resztą nawet zablokowanie możliwości dodania nr przesyłki i tak nie sprawi, że ktoś wyśle towar listem zwykły, a nie poleconym w omawianej sytuacji.

**Comment by MartaNowaczyk** [comment]
@vendoremo jeśli kupujący wybrał list ekonomiczny lub priorytetowy, to przez REST API nie powinno być problemu z dodaniem do takiego zamówienia numeru listu przewozowego.

**Comment by vendoremo** [comment]
No dobra, to może coś źle robię? Tym razem dodanie nr przesyłki już normalnie, dla wybranej formy dostawy DHL.

URI https://api.allegro.pl/order/checkout-forms/ZZZZZZZZ/shipments
Request json
string(135) "{"carrierId":"DHL","waybill":"20334424052","lineItems":["XXXXXXX","YYYYYYY"]}"
Response code
int(415)
Response message
array(1) { ["errors"]=> array(1) { [0]=> array(5) { ["code"]=> string(21) "NotSupportedException" ["message"]=> string(21) "An error has occurred" ["details"]=> NULL ["path"]=> NULL ["userMessage"]=> string(17) "An error occurred" } } }

**Comment by jakubmaguza** [comment]
Zgubiłeś ID w LineItems:
"lineItems": [
        {
        "id": "31896450-2de1-11e9-bced-352ab82ad9cd"  
        },
        {
        "id": "5b1f4a62-6969-11e8-b090-8b2d9f391430" 
        }
    ]

**Comment by vendoremo** [comment]
Dziękuję!

**Comment by vendoremo** [comment]
No dobra, ale dalej nie działa. Błąd bez zmian.

Json request:
{"carrierId":"DHL","waybill":"20334424052","lineItems":[{"id":"15313350-5145-11e9-a9f3-b32ec7a30bb2"},{"id":"15313351-5145-11e9-a9f3-b32ec7a30bb2"}]}

**Comment by jakubmaguza** [comment]
Hmm... Wysyłasz to POSTem?
Ja bym najpierw dodał tylko jedno ID w lineitems dla testu.
Sprawdziłbym czy GUID w urlu to faktycznie guid tego checkout formsa.
Potem sprawdził czy GUID z LineID faktycznie się zgadza (one są bardzo podobne do siebie, ale się różnią!)

No i oczywiście wszystkie pozostałe nagłówki w wersji beta :)

**Comment by vendoremo** [comment]
Tak, wysyłam POSTem. Skończyły mi się pomysły...., wklejam cały kod. Wiem, że GUIDy są prawie identyczne. Tak czy inaczej jeżeli podam zły GUID to powinien być inny kod błędu. A tu taki błąd nie-wiadomo-co.

<?php

$headers = array
(
	'Accept: application/vnd.allegro.beta.v1+json',
	'Content-Type: application/json',
	'Authorization: Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
);

$uri = 'https://api.allegro.pl/order/checkout-forms/15313353-5145-11e9-a9f3-b32ec7a30bb2/shipments';

$data = array (
	'carrierId' => 'DHL',
	'waybill'  => '20334424052',
	'lineItems' => array(array('id'=>'15313350-5145-11e9-a9f3-b32ec7a30bb2'), array('id'=>'15313351-5145-11e9-a9f3-b32ec7a30bb2'))
);

$session = curl_init($uri);

$requestjson = json_encode($data, JSON_UNESCAPED_UNICODE);

$session  = curl_init($uri);
curl_setopt($session, CURLOPT_HTTPHEADER, $headers);
curl_setopt($session, CURLOPT_POST, true);
curl_setopt($session, CURLOPT_POSTFIELDS, $requestjson);
curl_setopt($session, CURLOPT_RETURNTRANSFER, true);

$response = curl_exec($session);
$httpCode = curl_getinfo($session, CURLINFO_HTTP_CODE);
curl_close($session);

$wyniki = json_decode($response,TRUE);

?>

**Comment by MartaNowaczyk** [comment]
@vendoremo powinno być 'Content-Type: application/vnd.allegro.beta.v1+json'

**Comment by vendoremo** [comment]
@MartaNowaczyk - dziękuję, działa. Czy planowany jest zasób umożliwiający oznaczenie zamówienia jako wysłane w zakładce zamówienia? Takie oznaczenie jest warunkiem uzyskania 5* za czas wysyłki i koszty dostawy, także to istotny temat. Nie widzę w dokumentacji nic na ten temat, a trochę szkoda czasu na przeklikiwanie tego.

**Comment by MartaNowaczyk** [comment]
@vandoremo nie, nie planujemy. Ta opcja nie jest brana pod uwagę przy 5*. To dodanie numeru przesyłki jest tutaj priorytetowe. Oznaczenie o którym piszesz pozwala sprzedającym, którzy korzystają tylko z zakładek allegro do zarządzania swoją sprzedażą, filtrować swoje zamówienia.

**Comment by romigrabarz** [comment]
Czy metoda dodawania numeru listu przewozowego działa na sandbox? Dostaję odpowiedź: BadRequest.
Używam kodu wygenerowanego przez swagger.yaml:

 ```
var queryParams = new List<KeyValuePair<string, string>>();
            var headerParams = new Dictionary<string, string>
            {
                { "Authorization", $"Bearer {token}" },
                { "Accept", "application/vnd.allegro.beta.v1+json" },
                { "Content-Type", "application/vnd.allegro.beta.v1+json" }
            };
            var formParams = new Dictionary<string, string>()
            {
                { "carrierId", kurier.ToUpper() },
                { "waybill", listPrzewozowyNumer },
                { "lineItems", $"[{{{pozycjeZamowieniaAllegro.ToDelimitedString("},{",x=>$"\"id\":\"{x}\"")}}}]" }
            };
            var fileParams = new Dictionary<string, RestSharp.FileParameter>();
            var pathParams = new Dictionary<string, string>();
            var contentType = "application/x-www-form-urlencoded";
            var response = Client.CallApi($"https://api.allegro.pl{(Konto.Sandbox ? ".allegrosandbox.pl" : string.Empty)}/order/checkout-forms/{checkoutFormId}/shipments", RestSharp.Method.POST, queryParams, null, headerParams, formParams, fileParams, pathParams, contentType);
            if (response.StatusCode != HttpStatusCode.OK)
            {
                var error = Client.Deserialize(response, typeof(ErrorsHolder)) as ErrorsHolder;
                throw new Exception(error.Errors.ToDelimitedString(Environment.NewLine, x => $"{x.Message}\n{x.UserMessage}"));
            }
```

Poniżej kod czystym RestSharpem:
```
var request = new RestRequest($"https://api.allegro.pl.allegrosandbox.pl/order/checkout-forms/6878b3b1-adfe-11e9-9777-e9eedbce0edb/shipments", RestSharp.Method.POST);
            request.AddHeader("Authorization", $"Bearer {token}");
            request.AddHeader("Accept", "application/vnd.allegro.beta.v1+json");
            request.AddHeader("Content-Type", "application/vnd.allegro.beta.v1+json");
            request.AddParameter("carrierId", "DPD");
            request.AddParameter("waybill", "0000000527246Q");
            request.AddParameter("lineItems", "[{\"id\":\"05307130-adfe-11e9-b536-fd073a5eaf67\"},{\"id\":\"47364c30-adfe-11e9-b536-fd073a5eaf67\"}]");
            var response = RestClient.Execute(request);
```

Co robię źle?

**Comment by PrzemyslawLukanowski** [comment]
@romigrabarz  Zweryfikowałem to i dodawanie numeru przesyłki na sandboxie działa prawidłowo. Czy możesz przesłać nam tutaj przykładowy request w formacie cURL?

**Comment by romigrabarz** [comment]
Już poprawiłem. Miałem prawdopodobnie błąd w podawanych parametrach body. Zmieniłem sposób wywołania requesta przez klasę OrderManagementApi i teraz działa poprawnie :)

**Comment by softsimple** [comment]
Mam problem z przesłaniem numeru przesyłki na środowisku sandbox - dostaję 404 z informację "Order not found".
Request:
POST /order/checkout-forms/895989D2-2BAB-11EA-93CF-09416EE750B0/shipments HTTP/1.1
Host: https://api.allegro.pl.allegrosandbox.pl
Authorization: Bearer {token}
Accept: application/vnd.allegro.public.v1+json
Content-Type: application/vnd.allegro.public.v1+json
  {
	"carrierId":"FEDEX",
	"waybill":"6231377309838",
	"lineItems":
	[
		{
			"id":"895989d0-2bab-11ea-93cf-09416ee750b0"
		}
	]
  }


Response:
{
    "code": "ERROR",
    "message": "Order not found",
    "details": null,
    "path": null,
    "userMessage": null
}

Proszę o informację co jest nie tak? Czy środowisko sandbox działa poprawnie?

**Comment by dszewczyk1** [comment]
Czy da się z poziomu API usunąć numer przesyłki? Nie widzę informacji na ten temat w dokumentacji. Jednocześnie widzę, że jest to możliwe w Moim Allegro.

**Comment by PrzemyslawLukanowski** [comment]
@dszewczyk1 Niestety, ale na tę chwilę przez API nie da się usunąć numeru przesyłki.

**Comment by dszewczyk1** [comment]
@PrzemyslawLukanowski dziękuję za szybką odpowiedź. Czy są jakieś plany na wdrożenie takiej funkcjonalności? Czasami istnieje potrzeba wygenerowania nowej etykiety (np. zmiana w adresie), a osoby obsługujące sklep nie są jednocześnie administratorami, żeby zrobić to w Moje Allegro.

**Comment by PrzemyslawLukanowski** [comment]
@dszewczyk1 Aktualnie nie planujemy dodawać tej funkcjonalności, jednak samą sugestię mamy już zapisaną na naszej liście. Z POST /order/checkout-forms/{orderID}/shipments w ramach jednego zamówienia możesz korzystać wielokrotnie, dzięki czemu wygenerujesz nowy, dodatkowy numer przesyłki.
